sharding/review/casper_v2.1.md

```diff
@@ -0,0 +1,362 @@
+    'total_attester_deposits': 'int64',
+    # Who attested
+    'attester_bitfield': 'bytes'
+}
+```
+
+Here's the crystallized state:
+
+```python
+fields = {
+    # List of active validators
+    'active_validators': [ValidatorRecord],
+    # List of joined but not yet inducted validators
+    'queued_validators': [ValidatorRecord],
+    # List of removed validators pending withdrawal
+    'exited_validators': [ValidatorRecord],
+    # The permutation of validators used to determine who
+    # cross-links what shard in this epoch
+    'current_epoch_shuffling': ['int24'],
+    # The current epoch
+    'current_epoch': 'int64',
+    # The last justified epoch
+    'last_justified_epoch': 'int64',
+    # The last finalized epoch
+    'last_finalized_epoch': 'int64',
+    # The current dynasty
+    'current_dynasty': 'int64',
+    # The next shard that cross-linking assignment will start from
+    'next_shard': 'int16',
+    # The current FFG checkpoint
+    'current_checkpoint': 'hash32',
+    # Records about the most recent crosslink `for each shard
+    'crosslink_records': [CrosslinkRecord],
+    # Total balance of deposits
+    'total_deposits': 'int256',
+    # Used to select the committees for each shard
+    'crosslink_seed': 'hash32',
+    # Last epoch the crosslink seed was reset
+    'crosslink_seed_last_reset': 'int64'
+}
+```
+
+Each ValidatorRecord is an object containing information about a validator:
+
+```python
+fields = {
+    # The validator's public key
+    'pubkey': 'int256',
+    # What shard the validator's balance will be sent to 
+    # after withdrawal
+    'withdrawal_shard': 'int16',
+    # And what address
+    'withdrawal_address': 'address',
+    # The validator's current RANDAO beacon commitment
+    'randao_commitment': 'hash32',
+    # Current balance
+    'balance': 'int64',
+    # Dynasty where the validator can
+    # (be inducted | be removed | withdraw their balance)
+    'switch_dynasty': 'int64'
+}
+```
+
+And a CrosslinkRecord contains information about the last fully formed crosslink to be submitted into the chain:
+
+```python
+fields = {
+    # What epoch the crosslink was submitted in
+    'epoch': 'int64',
+    # The block hash
+    'hash': 'hash32'
+}
+```
+
+### Beacon chain processing
+
+Processing the beacon chain is fundamentally similar to processing a PoW chain in many respects. Clients download and process blocks, and maintain a view of what is the current "canonical chain", terminating at the current "head". However, because of the beacon chain's relationship with the existing PoW chain, and because it is a PoS chain, there are differences.
+
+For a block on the beacon chain to be processed by a node, three conditions have to be met:
+
+* The parent pointed to by the `parent_hash` has already been processed and accepted
+* The main chain block pointed to by the `main_chain_ref` has already been processed and accepted
```

Comment:
Is there any in-protocol condition of `main_chain_ref`? e.g., the main chain block number should within `head_main_chain_block_number - X`?